incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -1026,49 +1106,95 @@ private void onKafkaData(
                     kafkaDataEx != null && kafkaDataEx.kind() == KafkaDataExFW.KIND_MERGED ? kafkaDataEx.merged() : null;
                 final KafkaKeyFW key = kafkaMergedDataEx != null ? kafkaMergedDataEx.key() : null;
 
-                fetchWill:
+                reactToSignal:
                 if (key != null)
                 {
                     if (payload == null)
                     {
+                        final OctetsFW typeHeaderNameOctets =
+                            new OctetsFW().wrap(TYPE_HEADER_NAME.value(), 0, TYPE_HEADER_NAME.length());
+                        final String type = kafkaMergedDataEx.headers().matchFirst(h -> h.name().equals(typeHeaderNameOctets))
+                            .value().get((b, o, m) -> b.getStringWithoutLengthUtf8(o, m - o));
+
+                        final String keyPostfix = type.equals(WILL_SIGNAL_NAME.asString()) ?
+                            WILL_SIGNAL_KEY_POSTFIX : EXPIRY_SIGNAL_KEY_POSTFIX;
+
                         final String clientId0 = key.value()
-                            .get((b, o, m) -> b.getStringWithoutLengthUtf8(o, m - o - WILL_SIGNAL_KEY_POSTFIX.length()));
-                        String16FW clientId = new String16FW(clientId0);
-                        if (willDeliverIds.containsKey(clientId))
+                            .get((b, o, m) -> b.getStringWithoutLengthUtf8(o, m - o - keyPostfix.length()));
+                        final String16FW clientId = new String16FW(clientId0);
```

Comment:
If the header values we need to compare to are `OctetsFW`, then we should be storing things in terms of `OctetsFW`, not `StringFW`, etc?